docs/mocksupport.rst

```diff
@@ -867,13 +738,338 @@ This adds the instances to the repository display of the previous example:
+
+    def post_register_setup(self, conn):
+        """Execute setup code after provider is registered."""
+        pass
+
+
+
+Creating method user-defined providers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The detailed requirements for implementing a user-defined method provider is
+in section :ref:`Default Method Provider`.
+
+The following is an example for defining a method provider.
+
+.. code-block:: python
+
+    import pywbem
+    import pywbem_mock
+
+    from pywbem_mock import MethodProvider, CIMParameter, CIMError, \
+        CIM_ERR_NOT_FOUND, CIM_ERR_METHOD_NOT_AVAILABLE
+
+    class Method1TestProvider(MethodProvider):
+
+        provider_clasname = 'CIM_Foo_sub_sub'
+        """
+        User method provider for InvokeMethod using CIM_Foo_sub_sub and method1.
+        """
+        def __init__(self, cimrepository):
+            self.cimrepository = cimrepository
+
+        def InvokeMethod(self, namespace, MethodName, ObjectName, Params):
+            """
+            The parameters and return for Invoke method are defined in
+            :meth:`~pywbem_mock.MethodProvider.InvokeMethod`
+
+            This acts as both a static (ObjectName is only classname) and
+            dynamic (ObjectName is an instance name) method provider.
+            """
+            # Validate namespace using method in BaseProvider
+            self.validate_namespace(namespace)
+
+            # Get classname and validate. This provider uses only one class
+            if isinstance(ObjectName, six.string_types):
+                classname = ObjectName
+            else:
+                classname = ObjectName.classname
+            assert classname.lower() == 'tst_class'
+
+            # Test if class exists.
+            if not self.class_exists(namespace, classname):
+                raise CIMError(
+                    CIM_ERR_NOT_FOUND,
+                    _format("class {0|A} does not exist in CIM repository, "
+                            "namespace {1!A}", classname, namespace))
+
+            if isinstance(ObjectName, CIMInstanceName):
+                instance_store = self.get_instance_store(namespace)
+                inst = self.find_instance(objectname, instance_store,
+                                          copy=False)
+                if inst is None:
+                    raise CIMError(
+                        CIM_ERR_NOT_FOUND,
+                        _format("Instance {0|A} does not exist in CIM repository, "
+                                "namespace {1!A}", ObjectName, namespace))
+
+            if MethodName.lower() == 'method1':
+
+                rtn_parameters = [CIMParameter('OutputParam1', 'string',
+                                               value=namespace)]
+                else:
+                    rtn_params = None
+
+                return (return_value, rtn_params)
+
+            else:
+                raise CIMError(CIM_ERR_METHOD_NOT_AVAILABLE)
+
+    def run_invokemethod():
```

Comment:
What is the purpose of showing this function here? If we need to show it at all, it should be shown in a section that describes how to set up the mock support.

Reply:
None.  Copied to much from test code.  Note that I also cleaned up that example.